### PR TITLE
Hide Metal drawables from Kotlin runtime

### PR DESCRIPTION
**Pull request by elijah-semyonov**
Explicitly manage lifetime of metal drawables by hiding them from Kotlin runtime.

Fixes a memory spike when quickly resizing metal layer as in a scenario reported in [the issue](https://github.com/JetBrains/compose-multiplatform/issues/4850)

Otherwise associated drawable pools are retained until next GC, which can happen after inadequate amount of drawables with new sizes are allocated

###
Before:
Screenshot 2024-06-06 at 13 33 46
###
After:
Screenshot 2024-06-07 at 10 04 18

## Testing

Resize the popup with `ComposeUIViewController` to different detents inside Demo `IosBugs/PopupStretching`
Requires either patch:
```
Index: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/ComposeSceneMediator.uikit.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/ComposeSceneMediator.uikit.kt b/compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/ComposeSceneMediator.uikit.kt
--- a/compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/ComposeSceneMediator.uikit.kt	(revision 39d436a845d0cbe9526a6c24933268aacf83b2eb)
+++ b/compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/ComposeSceneMediator.uikit.kt	(date 1717747754590)
@@ -558,6 +558,7 @@
                 asDpRect().toRect()
             }
         }
+        onComposeSceneInvalidate()
         scene.size = IntSize(
             width = boundsInPx.width.roundToInt(),
             height = boundsInPx.height.roundToInt()
```
Or https://github.com/JetBrains/compose-multiplatform-core/pull/1387

### Fixes - iOS
- Fixed a memory spike when continuously resizing the `ComposeUIViewController` (such as when used in modal sheet presentation context with different detents)

### CODE REVIEW

File: compose/mpp/demo/src/uikitMain/kotlin/androidx/compose/mpp/demo/bugs/PopupStretching.kt [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.interop.LocalUIViewController
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.window.ComposeUIViewController
+import platform.UIKit.UIModalPresentationFormSheet
+import platform.UIKit.UISheetPresentationControllerDetent
+import platform.UIKit.sheetPresentationController
+
+//val bottomSheetUIViewController =
```
Comment (ASalavei) [inline compose/mpp/demo/src/uikitMain/kotlin/androidx/compose/mpp/demo/bugs/PopupStretching.kt]: Comments?

File: compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMetalDrawablesHandler.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+#import "CMPMacros.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface CMPMetalDrawablesHandler : NSObject
+
+- (instancetype)initWithMetalLayer:(CAMetalLayer *)metalLayer;
+
+- (void * CMP_OWNED)nextDrawable;
```
Comment (ASalavei) [inline compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMetalDrawablesHandler.h:44]: Just thinking about slightly more generic solution.
What do you think about creating analog of "Manual Reference Counter" that can wrap various classes and hide them from Kotlin GC?
Or maybe even play a bit with NSProxy to provide sufficient interface for Kotlin.

File: compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMetalDrawablesHandler.h [hunk 3]
```diff
@@ -0,0 +1,29 @@
+//
```
Comment (ASalavei) [inline compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMetalDrawablesHandler.h]: please check copyright

File: compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMetalDrawablesHandler.h (hunk 2 above)
Comment (elijah-semyonov) [reply compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMetalDrawablesHandler.h:44]: I tried to model this approach in my head, but didn't succeed. Do you have proof-of-concept available?

File: compose/mpp/demo/src/uikitMain/kotlin/androidx/compose/mpp/demo/bugs/PopupStretching.kt (hunk 1 above)
Comment (elijah-semyonov) [reply compose/mpp/demo/src/uikitMain/kotlin/androidx/compose/mpp/demo/bugs/PopupStretching.kt]: Oops, removed

File: compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMetalDrawablesHandler.h (hunk 3 above)
Comment (elijah-semyonov) [reply compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMetalDrawablesHandler.h]: Fixed!

File: compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMetalDrawablesHandler.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+#import <Foundation/Foundation.h>
+#import <QuartzCore/QuartzCore.h>
+#import <Metal/Metal.h>
+
+#import "CMPMacros.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface CMPMetalDrawablesHandler : NSObject
```
Comment (MatkovIvan) [inline compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMetalDrawablesHandler.h:36]: It requires a comment with explicit explanation why it's required

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/MetalRedrawer.uikit.kt [hunk 5]
```diff
@@ -372,7 +375,9 @@ internal class MetalRedrawer(
                     commandBuffer.label = "Present"
 
                     if (!presentsWithTransaction) {
-                        commandBuffer.presentDrawable(metalDrawable)
+                        // scheduleDrawablePresentation consumes metalDrawable
+                        // don't use metalDrawable after this call
```
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/MetalRedrawer.uikit.kt:385]: Can we just nullify it?

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/MetalRedrawer.uikit.kt [hunk 6]
```diff
@@ -145,6 +146,7 @@ internal class MetalRedrawer(
     private val metalLayer: CAMetalLayer,
     private val callbacks: MetalRedrawerCallbacks
 ) {
+    private val metalDrawableHandler = CMPMetalDrawablesHandler(metalLayer)
```
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/MetalRedrawer.uikit.kt]: Please add a note with a reason here too

File: compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMetalDrawablesHandler.h (hunk 4 above)
Comment (elijah-semyonov) [reply compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMetalDrawablesHandler.h:36]: What exactly? The class itself?

Comment (MatkovIvan) [reply compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMetalDrawablesHandler.h:36]: Yep. The reason why it's exist - why we need to move it outside K/N

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/MetalRedrawer.uikit.kt (hunk 5 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/MetalRedrawer.uikit.kt:385]: It's a smart cast `val`, doing so will require to put force-unwraps everywhere

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/MetalRedrawer.uikit.kt (hunk 6 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/MetalRedrawer.uikit.kt:149]: Got it.

Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/MetalRedrawer.uikit.kt]: Done

File: compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMetalDrawablesHandler.h (hunk 4 above)
Comment (elijah-semyonov) [reply compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMetalDrawablesHandler.h:36]: Done

**Review by MatkovIvan — APPROVED** [review]

**Review by ASalavei — APPROVED** [review]

**Comment by elijah-semyonov** [comment]
The particular web tests were reported to be flaky and are unrelevant to the PR, merging ignoring the failure.
Screenshot 2024-06-10 at 13 23 11

**Comment by chrisjenx** [comment]
Do you reckon this is related? https://github.com/JetBrains/compose-multiplatform/issues/4852

**Comment by elijah-semyonov** [comment]
Yes, sorry, your comment notification slipped from my attention. I am glad it has fixed your problem too.

**Pull request merged**